File: Backend/Travle.Model/Responses/TourScheduleResponse.cs
namespace Travle.Model.Responses
{
    /// <summary>
    /// A single date/time slot of a tour with its live seat picture. <see cref="FreeSeats"/> is the
    /// derived <c>Capacity - SeatsTaken</c> the mobile app shows travelers; <see cref="SeatsTaken"/> is
    /// maintained transactionally by the Phase-5 booking capacity guard, so this value becomes "live"
    /// the moment bookings start writing it. The <see cref="IsCancellable"/>/<see cref="IsDeletable"/>
    /// flags let the organizer desktop render slot actions as disabled-with-reason (rule K).
    /// </summary>
    public class TourScheduleResponse
    {
        public int Id { get; set; }
        public int TourId { get; set; }

        public DateTime StartsAt { get; set; }
        public DateTime EndsAt { get; set; }

        /// <summary>
        /// IANA zone (e.g. "Europe/Sarajevo") <see cref="StartsAt"/>/<see cref="EndsAt"/> display in — the
        /// tour's ordered-first destination's city zone. Those are UTC instants; the client converts them
        /// to this zone for display, labelled "(local time)". See docs/time-and-timezones.md.
        /// </summary>
        public string TimeZoneId { get; set; } = string.Empty;

        public int Capacity { get; set; }
        public int SeatsTaken { get; set; }

        /// <summary>Live free seats (<c>Capacity - SeatsTaken</c>), never below zero.</summary>
        public int FreeSeats { get; set; }

        /// <summary>
        /// Every booking row ever made on this slot — unlike <see cref="SeatsTaken"/>, which counts only
        /// bookings currently holding seats, this includes cancelled and expired ones. It is what blocks a
        /// hard delete (those rows are audit evidence), so it is the figure behind
        /// <see cref="IsDeletable"/> and <see cref="DeleteBlockedReason"/>.
        /// </summary>
        public int BookingCount { get; set; }

        /// <summary>Active / Cancelled — the enum name, never the raw int.</summary>
        public string Status { get; set; } = string.Empty;
        public string? CancelledReason { get; set; }
        public DateTime? CancelledAt { get; set; }

        /// <summary>
        /// When this departure stops accepting new bookings and completed payments: <see cref="StartsAt"/>
        /// minus the tour's booking cutoff (or the platform default). A UTC instant, displayed in
        /// <see cref="TimeZoneId"/> like the other event times.
        /// </summary>
        public DateTime BookingClosesAt { get; set; }

        /// <summary>
        /// True when a traveler can still book this slot right now: Active, before
        /// <see cref="BookingClosesAt"/>, and with a free seat. The clients gate their Book action on this
        /// rather than re-deriving the rule, so what they offer always matches what the server accepts.
        /// </summary>
        public bool IsBookable { get; set; }

        /// <summary>True when the slot is Active and still in the future (an organizer may cancel it).</summary>
        public bool IsCancellable { get; set; }

        /// <summary>
        /// True when the slot is Active, in the future and carries no booking rows at all — the exact
        /// condition <c>TourService.DeleteScheduleAsync</c> enforces, including cancelled and expired
        /// bookings, so the organizer desktop never offers a delete the server will refuse.
        /// </summary>
        public bool IsDeletable { get; set; }

        /// <summary>
        /// Why <see cref="IsDeletable"/> is false, for the desktop's disabled-with-reason tooltip (rule K).
        /// Null when the slot can be deleted.
        /// </summary>
        public string? DeleteBlockedReason { get; set; }

        public DateTime CreatedAt { get; set; }
        public DateTime? ModifiedAt { get; set; }
    }
}

File: UI/travle_core/lib/src/models/tour_schedule_response.dart
import 'package:json_annotation/json_annotation.dart';

part 'tour_schedule_response.g.dart';

/// A single date/time slot of a tour with its live seat picture (mirrors the
/// backend `TourScheduleResponse`). [freeSeats] is the derived `Capacity -
/// SeatsTaken`; the [isBookable]/[isCancellable]/[isDeletable] flags are the
/// server's own capability verdicts, so the apps gate their actions on them
/// rather than re-deriving the rules and drifting out of step.
@JsonSerializable()
class TourScheduleResponse {
  TourScheduleResponse({
    required this.id,
    required this.tourId,
    required this.startsAt,
    required this.endsAt,
    this.timeZoneId = 'UTC',
    required this.capacity,
    required this.seatsTaken,
    required this.freeSeats,
    required this.bookingCount,
    required this.status,
    required this.bookingClosesAt,
    required this.isBookable,
    required this.isCancellable,
    required this.isDeletable,
    required this.createdAt,
    this.cancelledReason,
    this.cancelledAt,
    this.deleteBlockedReason,
    this.modifiedAt,
  });

  final int id;
  final int tourId;

  final DateTime startsAt;
  final DateTime endsAt;

  /// IANA time-zone id (e.g. "Europe/Sarajevo") [startsAt]/[endsAt] display in — those are UTC instants.
  @JsonKey(defaultValue: 'UTC')
  final String timeZoneId;

  final int capacity;
  final int seatsTaken;
  final int freeSeats;

  /// Every booking row ever made on this slot, cancelled and expired ones
  /// included — unlike [seatsTaken], which counts only bookings still holding
  /// seats. This is what blocks a hard delete.
  final int bookingCount;

  /// Active / Cancelled — the enum name, never a raw int.
  final String status;
  final String? cancelledReason;
  final DateTime? cancelledAt;

  /// When this departure stops accepting new bookings and completed payments
  /// ([startsAt] minus the tour's booking cutoff). A UTC instant, displayed in
  /// [timeZoneId] like the other event times.
  final DateTime bookingClosesAt;

  /// Whether a traveler can book this slot right now — open, before
  /// [bookingClosesAt], and with a free seat.
  final bool isBookable;

  final bool isCancellable;
  final bool isDeletable;

  /// Why [isDeletable] is false, for the console's disabled-with-reason tooltip.
  final String? deleteBlockedReason;

  final DateTime createdAt;
  final DateTime? modifiedAt;

  bool get isActive => status == 'Active';
  bool get isCancelled => status == 'Cancelled';

  factory TourScheduleResponse.fromJson(Map<String, dynamic> json) =>
      _$TourScheduleResponseFromJson(json);

  Map<String, dynamic> toJson() => _$TourScheduleResponseToJson(this);
}
